src/HonzaBotner.Discord.Services/EventHandlers/BadgeRoleHandler.cs

```diff
@@ -0,0 +1,116 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using DSharpPlus.Entities;
+using DSharpPlus.EventArgs;
+using DSharpPlus.Exceptions;
+using HonzaBotner.Discord.EventHandler;
+using HonzaBotner.Discord.Services.Options;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace HonzaBotner.Discord.Services.EventHandlers;
+
+public class BadgeRoleHandler : IEventHandler<GuildMemberUpdateEventArgs>
+{
+    private readonly ILogger<BadgeRoleHandler> _logger;
+    private readonly BadgeRoleOptions _options;
+
+    public BadgeRoleHandler(IOptions<BadgeRoleOptions> options, ILogger<BadgeRoleHandler> logger)
+    {
+        _options = options.Value;
+        _logger = logger;
+    }
+
+    public async Task<EventHandlerResult> Handle(GuildMemberUpdateEventArgs args)
+    {
+        await Task.Delay(0); // Avoid warning that there is no "await" in async method
+        if (args.RolesAfter.Count > args.RolesBefore.Count) _ = Task.Run(() => CheckAddedRoles(args));
+        else if (args.RolesAfter.Count < args.RolesBefore.Count) _ = Task.Run(() => CheckRemovedRoles(args));
+        return EventHandlerResult.Continue;
+    }
+
+    private async Task CheckAddedRoles(GuildMemberUpdateEventArgs args)
+    {
+        DiscordRole addedRole = args.RolesAfter.First(role => !args.RolesBefore.Contains(role));
+
+        // Has the new role some badge and has user some of the trigger roles?
+        if (_options.PairedRoles.ContainsKey(addedRole.Id.ToString())
+            && args.RolesBefore.Select(role => role.Id).Any(roleId => _options.TriggerRoles.Contains(roleId)))
+        {
+            DiscordRole grantedRole = args.Guild.GetRole(_options.PairedRoles[addedRole.Id.ToString()]);
+            try
+            {
+                await args.Member.GrantRoleAsync(grantedRole);
+            }
+            catch (Exception e)
+            {
+                _logger.LogWarning(e, "Failed while assigning single role {RoleName} to {MemberName}",
+                    grantedRole.Name, args.Member.DisplayName);
+            }
```

Comment:
The `grantRole =` line might throw an exception if the role in the config is wrong.